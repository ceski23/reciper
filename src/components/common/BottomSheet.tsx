/* eslint-disable @typescript-eslint/no-use-before-define */
import styled from '@emotion/styled';
import { animated, useSpring } from '@react-spring/web';
import { useDrag } from '@use-gesture/react';
import {
  FC, useCallback, useEffect, useState,
} from 'react';

import { ReactPortal } from 'components/common/modal/ReactPortal';

import { exhaustiveCheck } from 'utils/guards';
import { color } from 'utils/styles/theme';

export type SheetState = 'open' | 'close' | 'peek';

interface BottomSheetProps {
  state: SheetState;
  onStateChange: (state: SheetState) => void;
  disableDrag?: boolean;
  gap?: number;
}

export const BottomSheet: FC<BottomSheetProps> = ({
  state, onStateChange, children, disableDrag, gap = 40,
}) => {
  const [sheetHeight, setSheetHeight] = useState(0);
  const [showBackdrop, setShowBackdrop] = useState(false);

  const measuredRef = useCallback((node: HTMLElement | null) => {
    if (node) setSheetHeight(node.getBoundingClientRect().height);
  }, []);

  const backdropStyles = useSpring({ opacity: state !== 'close' ? 1 : 0 });
  const [{ y }, api] = useSpring({
    y: 0,
    onStart(result) {
      if (result.value.y !== 0) setShowBackdrop(true);
    },
    onRest(result) {
      if (result.value.y === 0) setShowBackdrop(false);
    },
  }, []);

  const openSheet = useCallback(() => {
    api.start({ y: -sheetHeight + gap, immediate: false });
    onStateChange('open');
  }, [api, gap, onStateChange, sheetHeight]);

  const closeSheet = useCallback(() => {
    api.start({ y: 0, immediate: false });
    onStateChange('close');
  }, [api, onStateChange]);

  const peekSheet = useCallback(() => {
    api.start({ y: -sheetHeight * 0.4, immediate: false });
    onStateChange('peek');
  }, [api, onStateChange, sheetHeight]);

  useEffect(() => {
    switch (state) {
      case 'open': openSheet(); break;
      case 'close': closeSheet(); break;
      case 'peek': peekSheet(); break;

      default:
        exhaustiveCheck(state, 'Not all states checked');
        break;
    }
  }, [closeSheet, openSheet, peekSheet, state]);

  const bind = useDrag(({
    velocity: [,vy], offset: [,my], direction: [,dy], last,
  }) => {
    if (disableDrag) return;

    // If finished dragging
    if (last) {
      // If dragged upwards then open sheet
      if (dy < 0) openSheet();
      // If dragged quickly downwards or opened less than 40% of size then close sheet
      else if ((dy > 0 && vy > 0.5) || my > -sheetHeight * 0.4) closeSheet();
      // Open sheet
      else openSheet();
    } else {
      api.start({ y: my, immediate: true });
    }
  }, {
    delay: true,
    from: () => [0, y.get()],
    filterTaps: true,
    bounds: {
      top: -sheetHeight,
    },
  });

  return (
    <ReactPortal wrapperId="filtersModal">
      {showBackdrop && (
        <Backdrop style={backdropStyles} onClick={() => onStateChange('close')} />
      )}

      <Container ref={measuredRef} {...bind()} style={{ y }}>
        <Handle />
        <Content>{children}</Content>
      </Container>
    </ReactPortal>

  );
};

const Container = styled(animated.div)`
  background-color: ${color('background')};
  position: fixed;
  bottom: -100vh;
  left: 0;
  right: 0;
  z-index: 99;
  height: 100vh;
  touch-action: none;
  border-radius: 10px 10px 0 0;
  overflow: hidden;
  display: flex;
  flex-direction: column;
  align-items: center;
  box-shadow: 0px 0px 20px ${color('shadow')};
`;

const Handle = styled.div`
  background-color: ${color('backgroundAltHover')};
  width: 40px;
  height: 4px;
  border-radius: 10px;
  margin: 20px;
`;

const Backdrop = styled(animated.div)`
  background-color: #00000075;
  position: fixed;
  inset: 0;
  width: 100%;
  height: 100%;
  cursor: pointer;
  z-index: 99;
`;

const Content = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  flex: 1;
  padding: 0 20px 20px;
`;
