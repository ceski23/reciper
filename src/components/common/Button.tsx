/** @jsxImportSource @emotion/react */
import { css, Theme } from '@emotion/react/macro';
import styled from '@emotion/styled/macro';
import { hsl, parseToHsl, transparentize } from 'polished';
import React, { FC } from 'react';
import { match } from 'ts-pattern';

export type ButtonVariant = 'normal' | 'primary';
export type ButtonSize = 'normal' | 'small';

export const buttonStyles = (
  { theme, variant, size }: { theme: Theme, variant: ButtonVariant, size: ButtonSize },
) => css`
  font-weight: bold;
  font-family: Raleway, sans-serif;
  border: none;
  transition: background-color 0.3s, color 0.3s;
  outline: none;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  width: auto;
  cursor: pointer;
  box-shadow: 0px 3px 2px #0000001c;
  
  &:disabled {
    opacity: 0.7;
    cursor: not-allowed;
  }

  ${match(variant)
    .with('normal', () => css`
      background-color: ${theme.colors.backgroundhover};
      color: ${theme.colors.textalt};

      &:disabled {
        background-color: ${theme.colors.background2};
      }

      &:not(:disabled) {
        &:hover,
        &:focus-visible {
          background-color: ${hsl({
    hue: parseToHsl(theme.colors.backgroundhover).hue,
    saturation: parseToHsl(theme.colors.backgroundhover).saturation,
    lightness: parseToHsl(theme.colors.backgroundhover).lightness - 0.1,
  })};
        }
      }
    `)
    .with('primary', () => css`
      background-color: ${theme.colors.primary};
      color: ${transparentize(0, theme.colors.textLight)};

      &:disabled {
        background-color: ${theme.colors.primaryDisabled};
        color: ${theme.colors.text};
      }

      &:not(:disabled) {
        &:hover,
        &:focus-visible {
          background-color: ${theme.colors.primaryHover};
        }
      }
    `)
    .exhaustive()}
  
  ${match(size)
    .with('normal', () => css`
      font-size: 18px;
      border-radius: 10px;
      padding: 15px 15px;
      line-height: 1;

      & > svg, & > img {
        margin-right: 15px;
        width: 20px;
        height: 20px;
      }

      & > .filler {
        margin-left: 15px;
        width: 20px;
        height: 20px;
      }
    `)
    .with('small', () => css`
      font-size: 14px;
      border-radius: 8px;
      padding: 12px 12px;
      line-height: 1;

      & > svg, & > img {
        margin-right: 15px;
        width: 15px;
        height: 15px;
      }

      & > .filler {
        margin-left: 15px;
        width: 15px;
        height: 15px;
      }
    `)
    .exhaustive()}
`;

export const StyledButton = styled.button<{
  variant: ButtonVariant,
  size: ButtonSize
}>(buttonStyles);

export type HTMLButtonProps = React.DetailedHTMLProps<
React.ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement
>;

interface Props {
  icon?: React.FunctionComponent<React.SVGProps<SVGSVGElement>> | string;
  variant?: ButtonVariant;
  size?: ButtonSize;
}

export const Button: FC<HTMLButtonProps & Props> = ({
  children, icon: Icon, variant = 'normal', size = 'normal', ...props
}) => (
  <StyledButton size={size} variant={variant} {...props}>
    {Icon && (typeof Icon === 'string' ? <img src={Icon} className="filler" alt="" /> : <Icon />)}
    {children}
    {Icon && <span className="filler" />}
  </StyledButton>
);
