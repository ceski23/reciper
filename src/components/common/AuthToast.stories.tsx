import { ComponentMeta, ComponentStory } from '@storybook/react';

import { AuthToast } from 'components/common/AuthToast';

export default {
  title: 'Components/AuthToast',
  component: AuthToast,
} as ComponentMeta<typeof AuthToast>;

const Template: ComponentStory<typeof AuthToast> = (args) => <AuthToast {...args} />;

export const Default = Template.bind({});
Default.args = {};
