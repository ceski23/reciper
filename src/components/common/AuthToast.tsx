import styled from '@emotion/styled';
import { VFC } from 'react';

import { Link } from 'components/common/Link';

import { urls } from 'routing/urls';

import { color } from 'utils/styles/theme';

export const CustomToastIcon = styled.img`
  width: 20px;
  height: 20px;
`;

const ToastContainer = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  width: 100%;
`;

const ToastLink = styled(Link)`
  margin-left: 20px;
  background-color: ${color('backgroundHover')};
  padding: 5px 10px;
  border-radius: 10px;
  color: ${color('textalt')};
  text-decoration: none;
  font-weight: 600;
`;

export const AuthToast: VFC = () => (
  <ToastContainer>
    <span>Wymagane ponowne uwierzytelnienie</span>
    <ToastLink to={urls.settings.account()}>OK</ToastLink>
  </ToastContainer>
);
