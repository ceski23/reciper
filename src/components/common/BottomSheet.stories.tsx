import { Meta, StoryObj } from '@storybook/react';
import { useState } from 'react';

import { BottomSheet, SheetState } from 'components/common/BottomSheet';

export default {
  title: 'Components/BottomSheet',
  component: BottomSheet,
} as ComponentMeta<typeof BottomSheet>;

const Template: ComponentStory<typeof BottomSheet> = ({ state, ...args }) => {
  const [sheetState, setSheetState] = useState<SheetState>(state);

  const handleStateChange = (newState: SheetState) => {
    args.onStateChange(newState);
    setSheetState(newState);
  };

  return (
    <BottomSheet {...args} state={sheetState} onStateChange={handleStateChange} />
  );
};

export const Default = Template.bind({});
Default.args = {
  state: 'peek',
};
