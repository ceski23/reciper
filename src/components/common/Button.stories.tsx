import { ComponentMeta, ComponentStory } from '@storybook/react';

import { ReactComponent as HomeIcon } from 'assets/common/home.svg';

import { Button } from 'components/common/Button';

export default {
  title: 'Components/Button',
  component: Button,
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

export const Normal = Template.bind({});
Normal.args = {
  variant: 'normal',
  children: 'Button',
};

export const Primary = Template.bind({});
Primary.args = {
  variant: 'primary',
  children: 'Button',
};

export const NormalSize = Template.bind({});
NormalSize.args = {
  size: 'normal',
  children: 'Button',
};

export const SmallSize = Template.bind({});
SmallSize.args = {
  size: 'small',
  children: 'Button',
};

export const WithIcon = Template.bind({});
WithIcon.args = {
  icon: HomeIcon,
  children: 'Button',
};
