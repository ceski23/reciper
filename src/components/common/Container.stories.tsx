import { ComponentStory } from '@storybook/react';

import { FluidContainer, FullBleed } from 'components/common/Container';

const loremIpsum = `
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. In accumsan luctus interdum. Fusce in efficitur sapien. Vestibulum volutpat ut mauris accumsan cursus. Aenean sit amet malesuada purus. Duis vel maximus arcu. Nam quis est tortor. Quisque elementum elit ac augue eleifend congue. Proin consectetur eget nisi at placerat. Nulla vel consectetur dolor, non feugiat ipsum. Interdum et malesuada fames ac ante ipsum primis in faucibus. Integer aliquam libero in viverra feugiat. Nulla facilisi. Cras pharetra efficitur nisi vitae tempus. Proin pellentesque ex cursus pellentesque hendrerit. Etiam ligula sem, pretium quis augue quis, aliquam feugiat lectus.
`;

export default {
  title: 'Components/Container',
};

export const FluidStory: ComponentStory<typeof FluidContainer> = (props) => (
  <FluidContainer {...props} />
);
FluidStory.args = { children: <p>{loremIpsum}</p> };
FluidStory.storyName = 'FluidContainer';

export const FullBleedStory: ComponentStory<typeof FullBleed> = (props) => (
  <FluidContainer>
    <FullBleed {...props} />
  </FluidContainer>
);
FullBleedStory.args = { children: <p>{loremIpsum}</p> };
FullBleedStory.storyName = 'FullBleed';
