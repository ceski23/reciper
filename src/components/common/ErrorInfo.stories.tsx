import { Meta, StoryObj } from '@storybook/react';

import { Button } from 'components/common/Button';
import { ErrorInfo } from 'components/common/ErrorInfo';

export default {
  title: 'Components/ErrorInfo',
  component: ErrorInfo,
} as ComponentMeta<typeof ErrorInfo>;

const Template: ComponentStory<typeof ErrorInfo> = (args) => <ErrorInfo {...args} />;

export const Default = Template.bind({});
Default.args = {
  error: 'Nie udało się pobrać danych',
};

export const WithActions = Template.bind({});
WithActions.args = {
  error: 'Nie udało się pobrać danych',
  actions: <Button>Spróbuj ponownie</Button>,
};
