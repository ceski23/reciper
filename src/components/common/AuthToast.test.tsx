import { renderWithProviders } from 'test/utils';

import { AuthToast } from 'components/common/AuthToast';

describe('AuthToast', () => {
  it('should render correctly', () => {
    const { asFragment } = renderWithProviders(<AuthToast />);

    expect(asFragment()).toMatchSnapshot();
  });
});
