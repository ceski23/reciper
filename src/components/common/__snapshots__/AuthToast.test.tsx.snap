// Vitest Snapshot v1

exports[`AuthToast > should render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  width: 100%;
}

.emotion-3 {
  -webkit-text-decoration: none;
  text-decoration: none;
  margin-left: 20px;
  background-color: #e7ecef;
  padding: 5px 10px;
  border-radius: 10px;
  color: #365463;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 600;
}

<div
    class="emotion-0 emotion-1"
  >
    <span>
      Wymagane ponowne uwierzytelnienie
    </span>
    <a
      class="emotion-2 emotion-3 emotion-4"
      href="/settings/account"
    >
      OK
    </a>
  </div>
</DocumentFragment>
`;
